README.md
# azstore

[![license](https://img.shields.io/badge/license-MIT-success.svg?style=flat-square)](https://github.com/Algure/azstore/blob/master/LICENSE)
[![pub package](https://img.shields.io/pub/v/azstore.svg?color=success&style=flat-square)](https://pub.dartlang.org/packages/azstore)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-success.svg?style=flat-square)](https://github.com/Algure/azstore/pulls)

Access azure storage options via REST APIs.

## Getting Started

This package handles all the encryption and formatting required to provide easy access to azure storage options via REST APIs in a flutter project.
The package currently provides functions to query and upload data to Azure blobs, tables and queues.

## 🎖 Installing

```yaml
dependencies:
  azstore: ^latest_version
```

### ⚡️ Import

```dart
import 'package:azstore/azstore.dart';
```

## 🎮 How To Use
> Get your connection string from the azure portal after simply creating a storage account (you can follow the walkthrough in the section [Creating Azure Storage Account](#creating-azure-storage-account))

### Azure Blob Functions.

Azure blob allows storage of unstructured data in containers.

Typical use cases are shown below.

```dart
Future<void> testUploadImage() async {
  File testFile =File('C:/Users/HP/Pictures/fdblack.png');
  Uint8List bytes = testFile.readAsBytesSync();
  var storage = AzureStorage.parse('your connection string');
  try {
    await storage.putBlob('/azpics/fdblack.png',
      bodyBytes: bytes,//Text can be uploaded to 'blob' in which case body parameter is specified instead of 'bodyBytes'
      contentType: 'image/png',
    );
  }catch(e){
    print('exception: $e');
  }
}
```

Delete blob operations can also be performed as shown.

```dart
Future<void> testDeleteBlob() async {
  var storage = AzureStorage.parse('your connection string');
  try {
    await storage.deleteBlob('/azpics/fdblack.png');
    print('done deleting');
  }catch(e){
    print('exception: $e');
  }
}
```

Also explore the `appendBlock`, `createContainer` and `deleteContainer` functions.

### Table Storage Functions

The Azure Table service offers structured NoSQL storage in the form of tables.
Tables can be managed using the `createTable`, `deleteTable` and `getTables` functions.
Table nodes/rows can be controlled using other functions such as `upsertTableRow`, `putTableRow`, `getTableRow` and `deleteTableRow`.

The following snippets show the use of some table access functions. Also refer to the [Azure Tables docs](https://docs.microsoft.com/en-us/rest/api/storageservices/payload-format-for-table-service-operations/) for allowed data types to insert in a table row.
The code documentation provides further help.

Use `upsertTableRow` when updating or adding new table row and `putTableRow` to replace or add a new row.

```dart
Future<void> testUpload2Table() async {
  var storage = AzureStorage.parse('your connection string');
  try {
    var myPartitionKey='partition_key';
    var myRowKey='237';
    Map<String, dynamic> rowMap={"Address":"Santa Clara",
    "Age":23,
    "AmountDue":200.23,
    "CustomerCode@odata.type":"Edm.Guid",
    "CustomerCode":"c9da6455-213d-42c9-9a79-3e9149a57833",
    "CustomerSince@odata.type":"Edm.DateTime",
     "CustomerSince":"2008-07-10T00:00:00",
     "IsActive":false,
     "NumberOfOrders@odata.type":"Edm.Int64",
     "NumberOfOrders":"255",
     "PartitionKey":"$myPartitionKey",
     "RowKey":"$myRowKey"
    };
    await storage.upsertTableRow(
        tableName: 'profiles',
        rowKey:myRowKey,
        bodyMap: rowMap
     );
  }catch(e){
    print('tables upsert exception: $e');
  }
}
```

Specific Table rows can be retrieved using `getTableRow` function as shown below. Filters can also be used to retrieve a list of table rows by using the `filterTableRows` function specifying the [filter logic](https://docs.microsoft.com/en-us/rest/api/storageservices/querying-tables-and-entities) in the `filter` parameter.

```dart
Future<void> testGetTableRow() async {
  var storage = AzureStorage.parse('your connection string');
  try {
    var myPartitionKey='partition_key';
    var myRowKey='unique_row_id';
    String result=await storage.getTableRow(
        tableName: 'profiles',
        partitionKey:myPartitionKey,
        rowKey:myRowKey,
        fields: ['Address','CustomerSince']
    );
    print('result: $result');
  }catch(e){
    print('tables get exception: $e');
  }
}

Future<void> testFilterTable() async {
  var storage = AzureStorage.parse('your connection string');
  List<dynamic> result=await storage.filterTableRows(tableName: 'profiles',
   filter: 'Age%20lt%2024', 
   fields: ['Age','CustomerSince'], 
   top:30 
   );
  print('\nFilter results');
  for(var res in result){
    print(res);
  }
}
```

Table rows can also be deleted.

```dart
Future<void> testDeleteTableRow() async {
  try {
    var storage = AzureStorage.parse('your connection string');
    await storage.deleteTableRow(tableName: 'profiles', partitionKey: 'fgtdssdas', rowKey: '232');
  }catch(e){
    print('delete exception: $e');
  }
}
```

### Azure Queue Functions

Azure Queue Storage allows you store large numbers of messages. Queues in your storage account can easily be managed using `createQueue`, `deleteQueue`, `getQList` and `getQData` functions.

```dart
Future<void> testCreateQ() async {
  try{
      var storage = AzureStorage.parse('your connection string');
      await storage.createQueue('myqueue');
  }catch(e){
     print('create queue error: $e');
  }
}

Future<void> testGetQData() async {
  var storage = AzureStorage.parse('your connection string');
  try {
    String qName='myqueue';
    Map<String, String> result = await storage.getQData(qName);
    print('showing $qName data:\n');
    for (var res in result.entries) {
      print('${res.key}: ${res.value}');
    }
  }catch(e){
    print('get queue data error: $e');
  }
}

Future<void> testGetQList() async {
  try{
  var storage = AzureStorage.parse('your connection string');
  List<String> result=await storage.getQList();

  print('showing queue list\n');
  for(String res in result){
    print(res);
  }
  }catch(e){
     print('get queue list error: $e');
  }
}

Future<void> testDeleteQ() async {
  try{
      var storage = AzureStorage.parse('your connection string');
      await storage.deleteQueue('myqueue');
  }catch(e){
    print('delete queue error: $e');
  }
}
```

To insert and access messages in a queue, the functions `putQMessage`, `getQmessages`, `peekQmessage`, `clearQmessages` and `delQmessages` can be used as shown below.

```dart
Future<void> testPutMessage() async {
  var storage = AzureStorage.parse('your connection string');
  try {
     await storage.putQMessage(qName:'ttable', message: 'testing queue updates');
  }catch(e){
    print('get data error: ${e.statusCode} ${e.message}');
  }
}

Future<void> testGetQMessages() async {
  var storage = AzureStorage.parse('your connection string');
  print('working on results...');
  try {
    List<AzureQMessage> result = await storage.getQmessages(qName: 'ttable',
      numOfmessages: 10
    );
    print('showing get results');
    for (var res in result) {
      print('message ${res}');
    }
  }catch (e){
    print('Q get messages exception $e');
  }
}

Future<void> testPeekQMessages() async {
  var storage = AzureStorage.parse('your connection string');
  try {
    List<AzureQMessage> result = await storage.peekQmessages(qName: 'ttable');
    print('showing peek results');
    for (var res in result) {
      print('message ${res.messageText}');
    }
  }catch (e){
    print('Q peek messages exception $e');
  }
}

Future<void> testClearQMessage() async {
  var storage = AzureStorage.parse('your connection string');
  try {
    await storage.clearQmessages('ttable');
  }catch(e){
    print('clear QM error: ${e.statusCode} ${e.message}');
  }
}

Future<void> testDeleteMessage() async {
  var storage = AzureStorage.parse('your connection string');
  try {
    await storage.delQmessage(qName: 'ttable',
     messageId: '27bc633b-4de0-42bf-bea6-0860bd410f4e',
      popReceipt: 'AgAAAAMAAAAAAAAAX3e0UwAg1wE='
      );
  }catch(e){
    print('delete QM error: $e');
  }
}
```
The package provides internal documentation and required function parameters to ease working with functions. Also refer to the [Azure official documentation](https://docs.microsoft.com/en-us/rest/api/storageservices/queue-service-rest-api) for details on queue operations and message lifecycle.

## Azstore Functions.

### Blob Functions.

- `createContainer`: Create new blob container.
- `deleteContainer`: Delete blob container.
- `putBlob`: Put/update blob value.
- `appendBlock`: Add block to blob.
- `deleteBlob`: Delete blob.

### Table Functions.

- `createTable`: Create azure table.
- `deleteTable`: Delete azure table.
- `getTables`: Get tables in storage account.
- `upsertTableRow`: Update/insert new table row/node.
- `putTableRow`: Insert/replace new table row/node.
- `getTableRow`: Get table row/node values.
- `filterTableRows`: Search through rows with specified filter.
- `deleteTableRow`: Delete table node/row.

### Queue Functions.

- `createQueue`: Create new queue.
- `getQData`: Get properties of a queue.
- `deleteQueue`: Delete queue from storage account.
- `getQList`: Get list of all queues.
- `putQMessage`: Insert message to queue.
- `getQmessages`: Get a list of queue message objects while changing visibilty.
- `peekQmessages`: Get a list of queue message objects without changing visibility.
- `delQmessage`: Delete queue message.
- `updateQmessage`: Update queue message.
- `clearQmessages`: Delete all queue messages.

## Creating Azure Storage Account

You would need an azure storage account to complete this walk through.

### STEP 1.
Navigate to your azure portal and create a resource by clicking the `create a resourse button` then select `storage accounts` or simply click the `storage accounts` button if it appears on your home page.

![createRes](https://user-images.githubusercontent.com/37802577/112473274-6d9f2f80-8d6e-11eb-92e7-1025c96023d5.png)


![selectstore](https://user-images.githubusercontent.com/37802577/112473279-6f68f300-8d6e-11eb-943d-4d8e912c75ca.png)

### STEP 2.
Enter details for your new storage account and then `Review + create` . Also feel free to explore options in the other tabs (`Networking`, `Data Protection`, `Advanced` and `Tags`) for more control over your storage account. (Review process may take a few seconds).

![storewalkthrough](https://user-images.githubusercontent.com/37802577/112473217-5d875000-8d6e-11eb-9a1b-c21735b6e8fc.png)

### STEP 3:
Complete account creation by clicking the `Create` button after review is complete.

![create](https://user-images.githubusercontent.com/37802577/112473256-6841e500-8d6e-11eb-8d68-4cf6bbb1842a.png)


### STEP 4:
Go to resource after deployment is complete.

![deploymentcompleted](https://user-images.githubusercontent.com/37802577/112473277-6ed05c80-8d6e-11eb-83fa-a01d5908adae.png).

### STEP 5:
In the resource page, navigate to the `Access keys` tab and `show keys`. The `show keys` button exposes your access keys and **connection string** which is all you need to use this flutter package.





CHANGELOG.md
## 1.1.0

Update dependencies to latest version.

## 1.0.0

Null safety. More functions.

## 0.0.4

Updated docs.

## 0.0.3

Updated docs, bug fixes and backward compatibility.

## 0.0.2

Updated docs.

## 0.0.1

Blob, tables and queue access






## 0.0.4

Updated docs.



lib/src/custom_multipart_request.dart

import 'dart:async';

import 'package:http/http.dart' as http;

class CustomMultipartRequest extends http.MultipartRequest {
  /// Creates a new [MultipartRequest].
  CustomMultipartRequest(
      String method,
      Uri url, {
        this.onProgress,
      }) : super(method, url);

  final void Function(int bytes, int totalBytes)? onProgress;

  /// Freezes all mutable fields and returns a single-subscription [ByteStream]
  /// that will emit the request body.
  http.ByteStream finalize() {
    final byteStream = super.finalize();
    if (onProgress == null) return byteStream;

    final total = this.contentLength;
    int bytes = 0;

    final t = StreamTransformer.fromHandlers(
      handleData: (List<int> data, EventSink<List<int>> sink) {
        bytes += data.length;
        print('$bytes ; $total ');
        if(onProgress != null) {
          onProgress!(bytes, total);
        }
        sink.add(data);
      },
    );
    final stream = byteStream.transform(t);
    return http.ByteStream(stream);
  }
}

example/lib/main.dart
import 'dart:io';
import 'dart:typed_data';

import 'package:azstore/azstore.dart';
import 'package:flutter/material.dart';
import 'package:flutter_image_compress/flutter_image_compress.dart';
import 'package:image_picker/image_picker.dart';
import 'package:path_provider/path_provider.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Azstore Demo App'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({required this.title, super.key});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final _picker = ImagePicker();
  final navTextStyle = const TextStyle(color: Colors.white, fontSize: 14);
  final _connectionString =
      'DefaultEndpointsProtocol=httxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx';
  bool _progress = false;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            if (_progress)
              const CircularProgressIndicator(),
            const Text('Push button to test function:'),
            ListView(
              shrinkWrap: true,
              children: <MyButton>[
                MyButton(
                  text: 'Upload table node',
                  buttonColor: Colors.blue,
                  onPressed: uploadTableNode,
                ),
                MyButton(
                  text: 'Delete table row',
                  buttonColor: Colors.red,
                  onPressed: deleteTableRow,
                ),
                MyButton(
                  text: 'Filter tables',
                  buttonColor: Colors.green,
                  onPressed: filterTable,
                ),
                MyButton(
                  text: 'Get table row',
                  onPressed: getTableRow,
                ),
                MyButton(
                  text: 'Put blob image',
                  buttonColor: Colors.blue,
                  onPressed: uploadBlobImage,
                ),
                MyButton(
                  text: 'Delete blob',
                  buttonColor: Colors.red,
                  onPressed: deleteBlob,
                ),
                MyButton(
                  text: 'Create Queue',
                  buttonColor: Colors.blue,
                  onPressed: createQ,
                ),
                MyButton(
                  text: 'Get Queue Data',
                  buttonColor: Colors.green,
                  onPressed: getQData,
                ),
                MyButton(
                  text: 'Insert queue message',
                  buttonColor: Colors.blue,
                  onPressed: putQMessage,
                ),
                MyButton(
                  text: 'Delete Queue',
                  buttonColor: Colors.red,
                  onPressed: deleteQ,
                ),
                MyButton(
                  text: 'Get Queue list',
                  buttonColor: Colors.green,
                  onPressed: getQList,
                ),
                MyButton(
                  text: 'Get all queue messages',
                  buttonColor: Colors.green,
                  onPressed: clearQMessage,
                ),
              ],
            )
          ],
        ),
      ),
    );
  }

  Future<void> deleteTableRow() async {
    try {
      var storage = AzureStorage.parse(_connectionString);
      await storage.deleteTableRow(
          tableName: 'profiles', partitionKey: 'fgtdssdas', rowKey: '232');
      // showInfoDialog(context, 'Delete Success');//Optional prompt
    } catch (e) {
      debugPrint('delete exception: $e');
      // showErrorDialog(context, e.toString());//ALTERNATIVE PROMPT
    }
  }

  Future<void> filterTable() async {
    var storage = AzureStorage.parse(_connectionString);
    debugPrint('working on results...');
    final result = await storage.filterTableRows(
      tableName: 'profiles',
      filter: 'Age%20lt%2024',
      fields: ['Age', 'CustomerSince', 'PartitionKey', 'RowKey'],
      top: 10,
    );
    debugPrint('showing filter results');
    for (String res in result) {
      debugPrint(res);
    }
    // showInfoDialog(context, 'Success');//Optional prompt
  }

  Future<void> uploadTableNode() async {
    var storage = AzureStorage.parse(_connectionString);
    try {
      var myPartitionKey = "sfsdfsrg57865746";
      var myRowKey = '237';
      Map<String, dynamic> rowMap = {
        "Address": "Santa Clara",
        "Age": 23,
        "AmountDue": 200.23,
        "CustomerCode@odata.0type": "Edm.Guid",
        "CustomerCode": "c9da6455-213d-42c9-9a79-3e9149a57833",
        "CustomerSince@odata.type": "Edm.DateTime",
        "CustomerSince": "2008-07-10T00:00:00",
        "IsActive": false,
        "NumberOfOrders@odata.type": "Edm.Int64",
        "NumberOfOrders": "255",
        "PartitionKey": myPartitionKey,
        "RowKey": myRowKey
      };
      await storage.upsertTableRow(
          tableName: 'profiles',
          rowKey: myRowKey,
          partitionKey: myPartitionKey,
          bodyMap: rowMap);
      // showInfoDialog(context, 'Success');//Optional prompt
    } catch (e) {
      debugPrint('tables upsert exception: $e');
    }
  }

  Future<void> getTableRow() async {
    try {
      var storage = AzureStorage.parse(_connectionString);
      var myPartitionKey = "fgtdss*das";
      var myRowKey = '232';
      String result = await storage.getTableRow(
          tableName: 'profiles',
          partitionKey: myPartitionKey,
          rowKey: myRowKey,
          fields: ['Address', 'CustomerSince']);
      debugPrint('result: $result');
      debugPrint('done');
      // showInfoDialog(context, 'Success');
    } catch (e) {
      debugPrint('tables get exception: $e');
    }
  }

  Future<void> deleteBlob() async {
    var storage = AzureStorage.parse('your connection string');
    try {
      await storage.deleteBlob('/azpics/fdblack.png');
      // showInfoDialog(context, 'Delete Success');//Optional prompt
    } catch (e) {
      debugPrint('exception: $e');
      // showErrorDialog(context, '$e');//Optional prompt
    }
  }

  Future<void> updateQMessage() async {
    var storage = AzureStorage.parse(_connectionString);
    debugPrint('working on results...');
    try {
      await storage.updateQmessage(
          qName: 'ttable',
          messageId: 'c9aaeea8-4d47-4cf8-a080-250fb218468f',
          popReceipt: 'AgAAAAMAAAAAAAAAzVPboAkg1wE=',
          message: 'testing update: This is an update');
      debugPrint('done');
      // showInfoDialog(context, ' Success');//Optional prompt
    } catch (e) {
      debugPrint('delete QM error: $e');
      // showErrorDialog(context, e.toString());//Optional prompt
    }
  }

  Future<void> createQ() async {
    var storage = AzureStorage.parse(_connectionString);
    await storage.createQueue('newer-queue');
    // showInfoDialog(context, 'Create success');//Optional prompt
  }

  Future<void> getQData() async {
    var storage = AzureStorage.parse(_connectionString);
    try {
      String qName = 'myqueue';
      Map<String, String> result = await storage.getQData(qName);
      debugPrint('showing $qName data:\n');
      for (var res in result.entries) {
        debugPrint('${res.key}: ${res.value}');
      }
      // showInfoDialog(context, 'Success');//Optional prompt
    } catch (e) {
      debugPrint('get data error: $e');
      // showErrorDialog(context, e.toString());
    }
  }

  Future<void> deleteQ() async {
    var storage = AzureStorage.parse(_connectionString);
    await storage.deleteQueue('newer-queue');
    debugPrint('done');
  }

  Future<void> getQList() async {
    var storage = AzureStorage.parse(_connectionString);
    List<String> result = await storage.getQList();
    debugPrint('showing queue list\n');
    for (String res in result) {
      debugPrint(res);
    }
    // showInfoDialog(context, 'Success');//Optional prompt
  }

  Future<void> putQMessage() async {
    var storage = AzureStorage.parse(_connectionString);
    debugPrint('working on results...');
    try {
      await storage.putQMessage(qName: 'ttable', message: 'testing expiration');
      if (mounted) showInfoDialog(context, 'Success'); //Optional prompt
    } catch (e) {
      debugPrint('get data error: $e');
      showErrorDialog(context, e.toString()); //Optional prompt
    }
  }

  Future<void> getQMessages() async {
    var storage = AzureStorage.parse(_connectionString);
    debugPrint('working on results...');
    try {
      List<AzureQMessage> result = await storage.getQmessages(
          qName: 'ttable', //Required
          numOfmessages:
              10 //Optional. Number of messages to retrieve. This package returns top 20 filter results when not specified.
          );
      debugPrint('showing results');
      for (var res in result) {
        debugPrint('message $res');
      }
      if (mounted) showInfoDialog(context, 'Success'); //Optional prompt
    } catch (e) {
      debugPrint('Q get messages exception $e');
      showErrorDialog(context, e.toString()); //Optional prompt
    }
  }

  Future<void> peekQMessages() async {
    var storage = AzureStorage.parse(_connectionString);
    try {
      List<AzureQMessage> result = await storage.peekQmessages(qName: 'ttable');
      debugPrint('showing peek results');
      for (var res in result) {
        debugPrint('message ${res.messageText}');
      }
      if (mounted) showInfoDialog(context, 'Success'); //Optional prompt
    } catch (e) {
      debugPrint('Q peek messages exception $e');
      showErrorDialog(context, e.toString()); //Optional prompt
    }
  }

  Future<void> clearQMessage() async {
    var storage = AzureStorage.parse(_connectionString);
    debugPrint('working on results...');
    try {
      await storage.clearQmessages('ttable');
      debugPrint('done');
      if (mounted) showInfoDialog(context, 'Success'); //Optional prompt
    } catch (e) {
      debugPrint('delete QM error: $e');
      showErrorDialog(context, e.toString()); //Optional prompt
    }
  }

  Future<void> deleteQMessage() async {
    var storage = AzureStorage.parse(_connectionString);
    debugPrint('working on results...');
    try {
      await storage.delQmessage(
          qName: 'ttable',
          messageId: '27bc633b-4de0-42bf-bea6-0860bd410f4e',
          popReceipt: 'AgAAAAMAAAAAAAAAX3e0UwAg1wE=');
      debugPrint('done');
      if (mounted) showInfoDialog(context, 'Success'); //Optional prompt
    } catch (e) {
      debugPrint('delete QM error: $e');
      showErrorDialog(context, e.toString()); //Optional prompt
    }
  }

  uploadBlobImage() async {
    try {
      showProgress(true);
      final tempFile = await _picker.pickImage(source: ImageSource.gallery);
      assert(tempFile != null);
      final compressedPath = await compressImage(
        File(tempFile != null ? tempFile.path : '').absolute.path,
      );
      if (compressedPath == null) return;
      Uint8List bytes = File(compressedPath).readAsBytesSync();
      var storage = AzureStorage.parse(_connectionString);
      var picId = getUniqueIdWPath(compressedPath);
      await storage.putBlob(
        '/gmart-pics/$picId.jpg',
        bodyBytes: bytes,
        contentType: 'image/png',
      );
      if (mounted) showInfoDialog(context, 'Success'); //Optional prompt
      showProgress(false);
    } catch (e) {
      showProgress(false);
      debugPrint('An error occured. Error: $e');
      showErrorDialog(context, e.toString()); //Optional prompt
    }
  }

  Future<String?> compressImage(String imagePath) async {
    final directory = await getApplicationDocumentsDirectory();
    String path = '${directory.path}/my_app_pics';
    if (!Directory(path).existsSync()) await Directory(path).create();
    String fileId = getUniqueIdWPath(imagePath);
    path += '/$fileId.jpg';
    File newFile = File(path);
    await newFile.create();
    final compressionFile = await FlutterImageCompress.compressAndGetFile(
      imagePath,
      path,
      quality: 25,
      rotate: 0,
    );
    return compressionFile?.path;
  }

  void showInfoDialog(BuildContext context, String text) {
    showCustomDialog(
        context: context, icon: Icons.info, iconColor: Colors.blue, text: text);
  }

  void showErrorDialog(BuildContext context, String errorText) {
    showCustomDialog(
      context: context,
      icon: Icons.warning,
      iconColor: Colors.red,
      text: errorText,
    );
  }

  void showCustomDialog({
    required BuildContext context,
    required IconData icon,
    required String text,
    Color? iconColor,
    List? buttonList,
  }) {
    List<Widget> butList = [];
    if (buttonList != null && buttonList.isNotEmpty) {
      for (var arr in buttonList) {
        butList.add(Expanded(
          child: Container(
            margin: const EdgeInsets.all(4.0),
            decoration: BoxDecoration(
              color: arr[1],
              borderRadius: BorderRadius.circular(20),
            ),
            child: TextButton(
                onPressed: arr[2],
                child: Padding(
                  padding: const EdgeInsets.all(2),
                  child: Text(
                    arr[0],
                    style: navTextStyle,
                  ),
                )),
          ),
        ));
      }
    }
    Dialog errorDialog = Dialog(
      shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(10)),
      backgroundColor: Colors.white,
      child: SizedBox(
        height: 350,
        child: Column(
          children: <Widget>[
            Expanded(
              child: Icon(
                icon,
                color: iconColor,
                size: 200,
              ),
            ),
            Padding(
              padding: const EdgeInsets.all(8.0),
              child: Text(
                text,
                style: TextStyle(
                    color: Colors.blue.shade900, fontWeight: FontWeight.bold),
                textAlign: TextAlign.center,
              ),
            ),
            const SizedBox(height: 20),
            Container(
              height: butList.isNotEmpty ? 50 : 2,
              padding: const EdgeInsets.all(8.0),
              child: Row(
                children: buttonList != null ? butList : [],
              ),
            )
          ],
        ),
      ),
    );
    showGeneralDialog(
        context: context,
        barrierLabel: text,
        barrierDismissible: true,
        barrierColor: Colors.black.withOpacity(0.5),
        transitionDuration: const Duration(milliseconds: 500),
        transitionBuilder: (_, anim, __, child) {
          return SlideTransition(
            position: Tween(begin: const Offset(-1, 0), end: const Offset(0, 0))
                .animate(anim),
            child: child,
          );
        },
        pageBuilder: (BuildContext context, _, __) => (errorDialog));
  }

  String getUniqueIdWPath(String path) {
    List unis = path.split('/');
    return unis[unis.length - 1].split('.')[0];
  }

  void showProgress(bool progress) {
    setState(() {
      _progress = progress;
    });
  }
}

class MyButton extends StatelessWidget {
  const MyButton({
    required this.text,
    this.buttonColor,
    this.textColor,
    this.onPressed,
    super.key,
  });

  final Color? buttonColor;
  final String text;
  final Function()? onPressed;
  final Color? textColor;

  @override
  Widget build(BuildContext context) {
    return Container(
      width: double.maxFinite,
      margin: const EdgeInsets.all(4),
      decoration: BoxDecoration(
          color: buttonColor ?? Colors.black,
          borderRadius: BorderRadius.circular(5)),
      child: TextButton(
        onPressed: onPressed ?? () {},
        child: Text(
          text,
          style: TextStyle(color: textColor ?? Colors.white),
        ),
      ),
    );
  }
}
